### DISCUSSION TITLE
Why are these RawContact deletions failing?

### DISCUSSION THREAD

**Discussion by yuanhoujun** [vestrel00/contacts-android#224] [Q&A] [answered]
Use this code to delete raw contacts, it returns true.
```
val contacts = Contacts(mContext)
val blankRawContacts = contacts.accounts().queryRawContacts().where { RawContactsFields.Id `in` request.ids }.find()
val rawContacts = blankRawContacts.mapNotNull { it.toRawContact(contacts) }
val isSuccessful = contacts.delete().rawContacts(rawContacts).commit().isSuccessful
```

Then use this code to query all raw contacts
```
   fun getAllContacts(context: Context): List<ContactBasicInfo> {
        val contacts =
            Contacts(context).query().orderBy(ContactsFields.DisplayNamePrimary.asc()).find()

        val rawContacts = mutableListOf<ContactBasicInfo>()

        contacts.forEach { contact ->
            contact.rawContacts.forEach { rawContact ->
                val contactBasicInfo = ContactBasicInfo(
                    id = rawContact.id,
                    contactId = contact.id,
                    displayNamePrimary = rawContact.name?.displayName,
                    phoneNumber = rawContact.phones.sortedByDescending { it.isSuperPrimary }
                        .sortedByDescending { it.isPrimary }
                        .map { it.number }
                        .joinToString(separator = ",")
                )
                rawContacts.add(contactBasicInfo)
            }
        }

        return rawContacts
    }
```

Returned results still contain the raw contacts deleted.

**Comment by vestrel00** [comment]
@yuanhoujun, this sounds like a duplicate of issue #172 that has been fixed since version [0.2.0](https://github.com/vestrel00/contacts-android/releases/tag/0.2.0). What version of the library are you using? If you are not using 0.2.0 or above, could you please upgrade and let me know if the issue persists?

**Comment by vestrel00** [comment]
@yuanhoujun, also looking at your code, I see that there are some improvements that can be made 😁 

Instead of,

```kotlin
fun getAllContacts(context: Context): List<ContactBasicInfo> {
        val contacts =
            Contacts(context).query().orderBy(ContactsFields.DisplayNamePrimary.asc()).find()

        val rawContacts = mutableListOf<ContactBasicInfo>()

        contacts.forEach { contact ->
            contact.rawContacts.forEach { rawContact ->
                val contactBasicInfo = ContactBasicInfo(
                    id = rawContact.id,
                    contactId = contact.id,
                    displayNamePrimary = rawContact.name?.displayName,
                    phoneNumber = rawContact.phones.sortedByDescending { it.isSuperPrimary }
                        .sortedByDescending { it.isPrimary }
                        .map { it.number }
                        .joinToString(separator = ",")
                )
                rawContacts.add(contactBasicInfo)
            }
        }

        return rawContacts
    }
```

You could just use a flat map,

```kotlin
fun getAllContacts(context: Context): List<ContactBasicInfo> = Contacts(context)
        .query()
        .orderBy(ContactsFields.DisplayNamePrimary.asc())
        .find()
        .flatMap { it.rawContacts }
        .map { rawContact ->
             ContactBasicInfo(
                    id = rawContact.id,
                    contactId = rawContact.contactId,
                    displayNamePrimary = rawContact.name?.displayName,
                    phoneNumber = rawContact.phones.sortedByDescending { it.isSuperPrimary }
                        .sortedByDescending { it.isPrimary }
                        .map { it.number }
                        .joinToString(separator = ",")
                )
        }
```

**Comment by yuanhoujun** [comment]
> @yuanhoujun, this sounds like a duplicate of issue #172 that has been fixed since version [0.2.0](https://github.com/vestrel00/contacts-android/releases/tag/0.2.0). What version of the library are you using? If you are not using 0.2.0 or above, could you please upgrade and let me know if the issue persists?

0.2.1

**Comment by yuanhoujun** [comment]
> @yuanhoujun, also looking at your code, I see that there are some improvements that can be made 😁
> 
> Instead of,
> 
> ```kotlin
> fun getAllContacts(context: Context): List<ContactBasicInfo> {
>         val contacts =
>             Contacts(context).query().orderBy(ContactsFields.DisplayNamePrimary.asc()).find()
> 
>         val rawContacts = mutableListOf<ContactBasicInfo>()
> 
>         contacts.forEach { contact ->
>             contact.rawContacts.forEach { rawContact ->
>                 val contactBasicInfo = ContactBasicInfo(
>                     id = rawContact.id,
>                     contactId = contact.id,
>                     displayNamePrimary = rawContact.name?.displayName,
>                     phoneNumber = rawContact.phones.sortedByDescending { it.isSuperPrimary }
>                         .sortedByDescending { it.isPrimary }
>                         .map { it.number }
>                         .joinToString(separator = ",")
>                 )
>                 rawContacts.add(contactBasicInfo)
>             }
>         }
> 
>         return rawContacts
>     }
> ```
> 
> You could just use a flat map,
> 
> ```kotlin
> fun getAllContacts(context: Context): List<ContactBasicInfo> = Contacts(context)
>         .query()
>         .orderBy(ContactsFields.DisplayNamePrimary.asc())
>         .find()
>         .flatMap { it.rawContacts }
>         .map { rawContact ->
>              ContactBasicInfo(
>                     id = rawContact.id,
>                     contactId = rawContact.contactId,
>                     displayNamePrimary = rawContact.name?.displayName,
>                     phoneNumber = rawContact.phones.sortedByDescending { it.isSuperPrimary }
>                         .sortedByDescending { it.isPrimary }
>                         .map { it.number }
>                         .joinToString(separator = ",")
>                 )
>         }
> ```

Yep, thank you for your advice.

**Comment by vestrel00** [comment]
> > @yuanhoujun, this sounds like a duplicate of issue #172 that has been fixed since version [0.2.0](https://github.com/vestrel00/contacts-android/releases/tag/0.2.0). What version of the library are you using? If you are not using 0.2.0 or above, could you please upgrade and let me know if the issue persists?
> 
> 0.2.1

I was unable to reproduce the issue on my end just now. I'll have to investigate more tomorrow. Could you please share some logs with me? When you create your `Contacts` instances, please create them like this;

```kotlin
 Contacts(context, logger = AndroidLogger(redactMessages = true))
```

Filter the LogCat for "ContactsDebug". Please copy-paste the log output here so I can see a bit more details 🙏

**Comment by yuanhoujun** [comment]
```
---------------------------- PROCESS ENDED (17615) for package com.youngfeng.android.assistant ----------------------------
---------------------------- PROCESS STARTED (7947) for package com.youngfeng.android.assistant ----------------------------
2022-06-14 13:38:53.409  7947-8090  ContactsDebug           com.youngfeng.android.assistant      D  Query {
                                                                                                        includeBlanks: true
                                                                                                        rawContactsWhere: null
                                                                                                        include: data2, data3, data1, data4, data5, data6, data7, data8, data9, data10, contact_id, lookup, display_name, display_name_alt, contact_last_updated_timestamp, photo_uri, photo_thumb_uri, photo_file_id, has_phone_number, _id, starred, custom_ringtone, send_to_voicemail, is_primary, is_super_primary, mimetype, data14, raw_contact_id
                                                                                                        where: null
                                                                                                        orderBy: display_name COLLATE NOCASE ASC
                                                                                                        limit: 2147483647
                                                                                                        offset: 0
                                                                                                        hasPermission: true
                                                                                                        isRedacted: true
                                                                                                    }
2022-06-14 13:38:53.981  7947-8090  ContactsDebug           com.youngfeng.android.assistant      D                     Query.Result {
                                                                                                        Number of contacts found: 545
                                                                                                        First contact: Contact(id=110, lookupKey=792i21048505621315840, rawContacts=[RawContact(id=110, contactId=110, addresses=[], emails=[], events=[], groupMemberships=[], ims=[], name=null, nickname=null, note=null, organization=Organization(id=283, rawContactId=110, contactId=110, isPrimary=false, isSuperPrimary=false, company=null, title=****, department=null, jobDescription=null, officeLocation=null, symbol=null, phoneticName=null, isRedacted=true), phones=[], photo=null, relations=[], sipAddress=null, websites=[], customDataEntities={}, isRedacted=true)], displayNamePrimary=****, displayNameAlt=****, lastUpdatedTimestamp=Wed Jan 05 09:51:49 GMT+08:00 2022, options=Options(id=110, starred=false, customRingtone=null, sendToVoicemail=false, isRedacted=true), photoFileId=null, photoUri=null, photoThumbnailUri=null, hasPhoneNumber=false, isRedacted=true)
                                                                                                        isRedacted: true
                                                                                                    } (538 milliseconds)
2022-06-14 13:39:24.077  7947-8195  ContactsDebug           com.youngfeng.android.assistant      D  AccountsRawContactsQuery {
                                                                                                        isProfile: false
                                                                                                        rawContactsWhere: null
                                                                                                        where: _id IN ('***')
                                                                                                        orderBy: _id COLLATE NOCASE ASC
                                                                                                        limit: 2147483647
                                                                                                        offset: 0
                                                                                                        hasPermission: true
                                                                                                        isRedacted: true
                                                                                                    }
2022-06-14 13:39:24.110  7947-8195  ContactsDebug           com.youngfeng.android.assistant      D                     AccountsRawContactsQuery.Result {
                                                                                                        Number of raw contacts found: 1
                                                                                                        First raw contact: BlankRawContact(id=110, contactId=110, displayNamePrimary=****, displayNameAlt=****, isRedacted=true)
                                                                                                        isRedacted: true
                                                                                                    } (33 milliseconds)
2022-06-14 13:39:24.114  7947-8195  ContactsDebug           com.youngfeng.android.assistant      D  Query {
                                                                                                        includeBlanks: true
                                                                                                        rawContactsWhere: null
                                                                                                        include: data2, data3, data1, data4, data5, data6, data7, data8, data9, data10, contact_id, lookup, display_name, display_name_alt, contact_last_updated_timestamp, photo_uri, photo_thumb_uri, photo_file_id, has_phone_number, _id, starred, custom_ringtone, send_to_voicemail, is_primary, is_super_primary, mimetype, data14, raw_contact_id
                                                                                                        where: raw_contact_id = '***'
                                                                                                        orderBy: _id COLLATE NOCASE ASC
                                                                                                        limit: 2147483647
                                                                                                        offset: 0
                                                                                                        hasPermission: true
                                                                                                        isRedacted: true
                                                                                                    }
2022-06-14 13:39:24.172  7947-8195  ContactsDebug           com.youngfeng.android.assistant      D                     Query.Result {
                                                                                                        Number of contacts found: 1
                                                                                                        First contact: Contact(id=110, lookupKey=792i21048505621315840, rawContacts=[RawContact(id=110, contactId=110, addresses=[], emails=[], events=[], groupMemberships=[], ims=[], name=null, nickname=null, note=null, organization=Organization(id=283, rawContactId=110, contactId=110, isPrimary=false, isSuperPrimary=false, company=null, title=****, department=null, jobDescription=null, officeLocation=null, symbol=null, phoneticName=null, isRedacted=true), phones=[], photo=null, relations=[], sipAddress=null, websites=[], customDataEntities={}, isRedacted=true)], displayNamePrimary=****, displayNameAlt=****, lastUpdatedTimestamp=Wed Jan 05 09:51:49 GMT+08:00 2022, options=Options(id=110, starred=false, customRingtone=null, sendToVoicemail=false, isRedacted=true), photoFileId=null, photoUri=null, photoThumbnailUri=null, hasPhoneNumber=false, isRedacted=true)
                                                                                                        isRedacted: true
                                                                                                    } (56 milliseconds)
2022-06-14 13:39:24.175  7947-8195  ContactsDebug           com.youngfeng.android.assistant      D  Delete {
                                                                                                        rawContactIds: [110]
                                                                                                        contactIds: []
                                                                                                        hasPermission: true
                                                                                                        isRedacted: true
                                                                                                    }
2022-06-14 13:39:24.194  7947-8195  ContactsDebug           com.youngfeng.android.assistant      D                     Delete.Result {
                                                                                                        isSuccessful: true
                                                                                                        rawContactIdsResultMap: {110=true}
                                                                                                        contactIdsResultMap: {}
                                                                                                        isRedacted: true
                                                                                                    } (20 milliseconds)
2022-06-14 13:39:26.858  7947-8195  ContactsDebug           com.youngfeng.android.assistant      D  Query {
                                                                                                        includeBlanks: true
                                                                                                        rawContactsWhere: null
                                                                                                        include: data2, data3, data1, data4, data5, data6, data7, data8, data9, data10, contact_id, lookup, display_name, display_name_alt, contact_last_updated_timestamp, photo_uri, photo_thumb_uri, photo_file_id, has_phone_number, _id, starred, custom_ringtone, send_to_voicemail, is_primary, is_super_primary, mimetype, data14, raw_contact_id
                                                                                                        where: null
                                                                                                        orderBy: display_name COLLATE NOCASE ASC
                                                                                                        limit: 2147483647
                                                                                                        offset: 0
                                                                                                        hasPermission: true
                                                                                                        isRedacted: true
                                                                                                    }
2022-06-14 13:39:27.218  7947-8195  ContactsDebug           com.youngfeng.android.assistant      D                     Query.Result {
                                                                                                        Number of contacts found: 545
                                                                                                        First contact: Contact(id=110, lookupKey=792i21048505621315840, rawContacts=[RawContact(id=110, contactId=110, addresses=[], emails=[], events=[], groupMemberships=[], ims=[], name=null, nickname=null, note=null, organization=Organization(id=283, rawContactId=110, contactId=110, isPrimary=false, isSuperPrimary=false, company=null, title=****, department=null, jobDescription=null, officeLocation=null, symbol=null, phoneticName=null, isRedacted=true), phones=[], photo=null, relations=[], sipAddress=null, websites=[], customDataEntities={}, isRedacted=true)], displayNamePrimary=****, displayNameAlt=****, lastUpdatedTimestamp=Wed Jan 05 09:51:49 GMT+08:00 2022, options=Options(id=110, starred=false, customRingtone=null, sendToVoicemail=false, isRedacted=true), photoFileId=null, photoUri=null, photoThumbnailUri=null, hasPhoneNumber=false, isRedacted=true)
                                                                                                        isRedacted: true
                                                                                                    } (340 milliseconds)

```

**Comment by vestrel00** [comment]
@yuanhoujun, I also just noticed something that can be improved by the library thanks to your code sample!

```kotlin
val contacts = Contacts(mContext)
val blankRawContacts = contacts.accounts().queryRawContacts().where { RawContactsFields.Id `in` request.ids }.find()
val rawContacts = blankRawContacts.mapNotNull { it.toRawContact(contacts) }
val isSuccessful = contacts.delete().rawContacts(rawContacts).commit().isSuccessful
```

I created an issue (#220) that will simplify this to,

```kotlin
val contacts = Contacts(mContext)
val blankRawContacts = contacts.accounts().queryRawContacts().where { RawContactsFields.Id `in` request.ids }.find()
val isSuccessful = contacts.delete().rawContacts(blankRawContacts).commit().isSuccessful
```

That will remove one database query per `blankRawContacts`, which is done every time `it.toRawContact(contacts)` is invoked.

I will also work on an existing issue (#158) that I have been planning to work on that will simplify this even more to,

```kotlin
val isSuccessful = Contacts(mContext).delete().rawContactsWhere { RawContact.Id `in` request.ids }.commit().isSuccessful
```

**UPDATE (June 14):** I also created another issue (#222), that does not require using WHERE. Usage would look like;

```kotlin
val isSuccessful = Contacts(mContext).delete().rawContacts(request.ids).commit().isSuccessful
```

**Comment by yuanhoujun** [comment]
Yep, I noticed #158, That'll be more natural and better performance. so when will you release this feature?

**Comment by vestrel00** [comment]
> Yep, I noticed #158, That'll be more natural and better performance. so when will you release this feature?
> 

Because you are interested in it, I will aim to have #158, #220, #222, and this bug you created by end of this week.

I will probably create another release just for these issues. So, it will be in 0.2.2 =)

**Comment by yuanhoujun** [comment]
Ok, wonderful, 👍, thanks for your hard work.

**Comment by vestrel00** [comment]
> ---------------------------- PROCESS STARTED (7947) for package com.youngfeng.android.assistant ----------------------------
> ... 

I'll look at these logs tomorrow. @yuanhoujun, could you also please provide me with even more logs? It would help me a lot to see the state of the RawContacts table before and after you commit the delete operation.

Please temporarily install the `debug` module;

```kotlin
 implementation 'com.github.vestrel00.contacts-android:debug:0.2.1'
```

Then, [log the RawContacts table](https://vestrel00.github.io/contacts-android/debug/debug-contacts-provider-tables/) before and after the delete operation. Also, please log the rawContactIds that should be deleted. So, something like;

```kotlin
Log.d("ContactsDebug", "DELETING RAW CONTACTS WITH IDS: ${request.ids}")
mContext.logRawContactsTable()

val contacts = Contacts(mContext)
val blankRawContacts = contacts.accounts().queryRawContacts().where { RawContactsFields.Id `in` request.ids }.find()
val rawContacts = blankRawContacts.mapNotNull { it.toRawContact(contacts) }
val isSuccessful = contacts.delete().rawContacts(rawContacts).commit().isSuccessful

mContext.logRawContactsTable()
```

Filter the LogCat with "ContactsDebug" and paste the output here. Please be careful NOT to paste sensitive personal information such as RawContact names. I advise you to use fake test data.

**Comment by yuanhoujun** [comment]
It's really personal, Could you add me on WeChat or Telegram? I can post these logs for you. I'm not sure this problem will appear if I use fake test data or another phone.

**Comment by yuanhoujun** [comment]
My telegram account: ouyangfeng2016, Wechat: ouyangfeng_office

**Comment by vestrel00** [comment]
> It's really personal, Could you add me on WeChat or Telegram? I can post these logs for you. I'm not sure this problem will appear if I use fake test data or another phone.
> My telegram account: ouyangfeng2016, Wechat: ouyangfeng_office

It's not necessary to WeChat or Telegram yet, but I may want to if I can't figure this out tomorrow.

I really only need to answer these questions.

1. Are the RawContacts represented by `request.ids` in the RawContacts table BEFORE the delete operation?
2. Are the RawContacts represented by `request.ids` still in the RawContacts table AFTER the delete operation?
    - If they are still there, what is the value of the `deleted` column for each of them?

Could you do what I suggested in https://github.com/vestrel00/contacts-android/issues/219#issuecomment-1154777614 but instead of pasting the entire RawContacts table, you would just look at it yourself and give me the answers to my two questions above? All you need to do is check if the RawContacts row with `id`s in `request.ids` are still there before and after the call to `commit()`.

**Comment by yuanhoujun** [comment]
Try this, I removed some sensitive words.
[contacts.log](https://github.com/vestrel00/contacts-android/files/8897984/contacts.log)

**Comment by vestrel00** [comment, marked as answer]
@yuanhoujun, based on the logs. I see that the RawContact you are attempting to delete is NOT getting deleted! It is also not even being marked for deletion. I think I know what's going on here.

I noticed that the RawContact you are trying to delete belongs to a `com.xiaomi` account. I think that this is a read-only RawContact; https://developer.android.com/reference/android/provider/ContactsContract.RawContactsColumns#RAW_CONTACT_IS_READ_ONLY

![Screen Shot 2022-06-14 at 7 29 59 AM](https://user-images.githubusercontent.com/4332753/173640163-003db744-8506-46fe-806c-3c956dc95302.png)

I was going to ask you to log the value of `ContactsContract.RawContacts.RAW_CONTACT_IS_READ_ONLY` in the RawContacts table but the query keeps throwing an exception when adding that column 😬 

If it is read-only, then we might not be able to delete it at all. The only way these RawContacts can be deleted is from the remote server/service that belongs to the `com.xiaomi` Account. You can also implement your own SyncAdapters, which is currently out of scope of this library. 

BUT before we come to any conclusions, **could you try deleting this Contact/RawContact using a different Contacts app?**

I'm not sure what Contacts apps are available in your region. I typically use the native AOSP Contacts app that comes pre-installed with most US Android phones. I also use;

- Google Contacts: https://play.google.com/store/apps/details?id=com.google.android.contacts
- Samsung Contacts: https://samsung-contacts.en.uptodown.com/android

Let me know if you are able to delete the Contact/RawContact using any of those apps (or similar apps) 🙏 

----------

I did verify that there is a bug in the delete APIs. The `Delete`, `ProfileDelete`, and `DataDelete` APIs can return success=true even if the delete actually failed. I created a separate issue for this; #223

The best that I might be able to do for this thread is to fix the issue ^ so that `val isSuccessful = contacts.delete().rawContacts(rawContacts).commit().isSuccessful` will return false instead of true. You will at least be able to tell your users that the contact could not be deleted.

 I am willing to help do more but I will wait until you try deleting Contact/RawContact using a different Contacts app.

I'll convert this issue to a discussion for now.

**Comment by yuanhoujun** [comment]
Your guess is right.
I tried another app from China, but it also failed when trying to delete contacts on "Xiaomi". 
So it's OK, you don't need to fix this issue. 

But I think you need to return false when failed, like #223, go ahead.
And add a read-only flag for RawContact and Contact.

**Comment by vestrel00** [comment]
> Your guess is right.
> 
> I tried another app from China, but it also failed when trying to delete contacts on "Xiaomi". 
> 
> So it's OK, you don't need to fix this issue. 

Thanks for checking! 

> But I think you need to return false when failed, like #223, go ahead.

Yes. I have already fixed that issue. I'm just working on other tickets I'd like to include in the next release (0.2.2). One of those issues will also benefit you (#222). The release should be public before end of Friday this week.

> And add a read-only flag for RawContact and Contact.

Unfortunately, I am unable to add a read-only flag. As I mentioned, including `ContactsContract.RawContacts.RAW_CONTACT_IS_READ_ONLY` in queries throws an exception. This is probably also why the Contacts app you are using allows you to try to delete the read-only RawContacts even though the delete will fail.

**Reply by yuanhoujun** [reply]
Nice job! @vestrel00

**Reply by vestrel00** [reply]
FYI, I'm planning to release 0.2.2 tomorrow around this time. Take a look at this discussion for what it will include; https://github.com/vestrel00/contacts-android/discussions/226

**Reply by vestrel00** [reply]
@yuanhoujun, I have released 0.2.2. Please take a look here; https://github.com/vestrel00/contacts-android/releases/tag/0.2.2

I suggest that you refactor your code from,

```kotlin
val contacts = Contacts(mContext)
val blankRawContacts = contacts.accounts().queryRawContacts().where { RawContactsFields.Id `in` request.ids }.find()
val rawContacts = blankRawContacts.mapNotNull { it.toRawContact(contacts) }
val isSuccessful = contacts.delete().rawContacts(rawContacts).commit().isSuccessful
```

To,

```kotlin
val isSuccessful = Contacts(mContext).delete().rawContactsWithId(request.ids).commit().isSuccessful
```

You can take a look at the documentation here; https://vestrel00.github.io/contacts-android/basics/delete-contacts/#a-basic-delete

Also, please keep in mind that `Delete.Result.isSuccessful` will return false even if just one of the RawContacts in `request.ids` failed to delete. It will only return true if ALL RawContacts in `request.ids` were deleted. If you want to check the success status of individual deletes, please do something like the following instead;

```kotlin
val deleteResult = Contacts(mContext).delete().rawContactsWithId(request.ids).commit()
for(rawContactId in request.ids) {
    val isSuccessful = deleteResult.isRawContactDeleteSuccessful(rawContactId)
    // Handle true or false
}
```

You can take a look at the documentation on result handling here; https://vestrel00.github.io/contacts-android/basics/delete-contacts/#handling-the-delete-result

Or, you can play around with `commitInOneTransaction()` and see if that suites your needs better than `commit()`,

```kotlin
val isSuccessful = Contacts(mContext).delete().rawContactsWithId(request.ids).commitInOneTransaction().isSuccessful
```

----------

Please let me know if you have any other issues or if something is not working correctly. I would be very happy to fix it! I thank you very much for participating in this discussion and contributing to the project. You are AWESOME 🔥 ❤️ 

Good luck and have fun! 😁

**Comment by vestrel00** [comment]
A new issue popped up that seems related to this; #293

**Reply by vestrel00** [reply]
It seems like this is a real issue with Xiaomi phones. So, I bought my own Xiaomi phone to debug! Watch issue #293 for more updates!

**Reply by vestrel00** [reply]
I confirmed that there is currently nothing we can do at this stage. @yuanhoujun is correct. I tried deleting Xiaomi contacts on my Xiaomi phone using 3rd party contacts apps and all of them failed to delete Xiaomi contacts.

More context in https://github.com/vestrel00/contacts-android/issues/293#issuecomment-1676573094
